File: src/main/kotlin/sunsetsatellite/lang/sunlite/Stmt.kt
package sunsetsatellite.lang.sunlite

abstract class Stmt: Element {

	data class Expression(val expr: Expr) : Stmt() {
		override fun <R> accept(visitor: Visitor<R>): R {
			return visitor.visitExprStmt(this)
		}

		override fun getLine(): Int {
			return expr.getLine()
		}

		override fun getFile(): String? {
			return expr.getFile()
		}
	}

	data class Print(val expr: Expr) : Stmt() {
		override fun <R> accept(visitor: Visitor<R>): R {
			return visitor.visitPrintStmt(this)
		}

		override fun getLine(): Int {
			return expr.getLine()
		}

		override fun getFile(): String? {
			return expr.getFile()
		}
	}

	data class Var(val name: Token, val type: Type, val initializer: Expr?, val modifier: FieldModifier) : Stmt(),
		NamedStmt {
		override fun <R> accept(visitor: Visitor<R>): R {
			return visitor.visitVarStmt(this)
		}

		override fun getLine(): Int {
			return name.line
		}

		override fun getFile(): String? {
			return name.file
		}

		override fun getNameToken(): Token {
			return name
		}
	}

	data class Block(val statements: List<Stmt>, val lineNumber: Int, val currentFile: String?) : Stmt() {
		override fun <R> accept(visitor: Visitor<R>): R {
			return visitor.visitBlockStmt(this)
		}

		override fun getLine(): Int {
			return lineNumber
		}

		override fun getFile(): String? {
			return currentFile
		}
	}

	data class If(val condition: Expr, val thenBranch: Stmt, val elseBranch: Stmt?) : Stmt() {
		override fun <R> accept(visitor: Visitor<R>): R {
			return visitor.visitIfStmt(this)
		}

		override fun getLine(): Int {
			return condition.getLine()
		}

		override fun getFile(): String? {
			return condition.getFile()
		}
	}

	data class While(val condition: Expr, val body: Stmt) : Stmt() {
		override fun <R> accept(visitor: Visitor<R>): R {
			return visitor.visitWhileStmt(this)
		}

		override fun getLine(): Int {
			return condition.getLine()
		}

		override fun getFile(): String? {
			return condition.getFile()
		}
	}

	class Break(val keyword: Token) : Stmt(){
		override fun <R> accept(visitor: Visitor<R>): R {
			return visitor.visitBreakStmt(this)
		}

		override fun getFile(): String? {
			return keyword.file
		}

		override fun getLine(): Int {
			return keyword.line
		}

	}

	class Continue(val keyword: Token) : Stmt(){
		override fun <R> accept(visitor: Visitor<R>): R {
			return visitor.visitContinueStmt(this)
		}

		override fun getFile(): String? {
			return keyword.file
		}

		override fun getLine(): Int {
			return keyword.line
		}
	}

	data class Function(val name: Token, val params: List<Param>, val body: List<Stmt>, var modifier: FunctionModifier, val returnType: Type, val typeParams: List<Param>) : Stmt(),
		NamedStmt {
		override fun <R> accept(visitor: Visitor<R>): R {
			return visitor.visitFunctionStmt(this)
		}

		override fun getLine(): Int {
			return name.line
		}

		override fun getFile(): String? {
			return name.file
		}

		override fun getNameToken(): Token {
			return name
		}
	}

	data class Return(val keyword: Token, val value: Expr?) : Stmt(){
		override fun <R> accept(visitor: Visitor<R>): R {
			return visitor.visitReturnStmt(this)
		}

		override fun getFile(): String? {
			return keyword.file
		}

		override fun getLine(): Int {
			return keyword.line
		}
	}

	data class Import(val keyword: Token, val what: Token) : Stmt(){
		override fun <R> accept(visitor: Visitor<R>): R {
			return visitor.visitImportStmt(this)
		}

		override fun getFile(): String? {
			return keyword.file
		}

		override fun getLine(): Int {
			return keyword.line
		}
	}

	data class Class(val name: Token, val methods: List<Function>, val fieldDefaults: List<Var>, val superclass: Expr.Variable?, val superinterfaces: List<Expr.Variable>, val modifier: ClassModifier, val typeParameters: List<Param>) : Stmt(),
		NamedStmt {
		override fun <R> accept(visitor: Visitor<R>): R {
			return visitor.visitClassStmt(this)
		}

		override fun getFile(): String? {
			return name.file
		}

		override fun getLine(): Int {
			return name.line
		}

		override fun getNameToken(): Token {
			return name
		}
	}

	data class Interface(val name: Token, val methods: List<Function>, val superinterfaces: List<Expr.Variable>, val typeParameters: List<Param>) : Stmt(),
		NamedStmt {
		override fun <R> accept(visitor: Visitor<R>): R {
			return visitor.visitInterfaceStmt(this)
		}

		override fun getFile(): String? {
			return name.file
		}

		override fun getLine(): Int {
			return name.line
		}

		override fun getNameToken(): Token {
			return name
		}
	}

	interface Visitor<R> {
		fun visitExprStmt(stmt: Expression): R
		fun visitPrintStmt(stmt: Print): R
		fun visitVarStmt(stmt: Var): R
		fun visitBlockStmt(stmt: Block): R
		fun visitIfStmt(stmt: If): R
		fun visitWhileStmt(stmt: While): R
		fun visitBreakStmt(stmt: Break): R
		fun visitContinueStmt(stmt: Continue): R
		fun visitFunctionStmt(stmt: Function): R
		fun visitReturnStmt(stmt: Return): R
		fun visitClassStmt(stmt: Class): R
		fun visitInterfaceStmt(stmt: Interface): R
		fun visitImportStmt(stmt: Import): R
	}

	interface NamedStmt: Element {
		fun getNameToken(): Token
	}

	abstract fun <R> accept(visitor: Visitor<R>): R
}

File: src/main/csharp/SunliteSharp/Core/AST/Stmt.cs
using SunliteSharp.Core.Enum;
using SunliteSharp.Core.Modifier;

namespace SunliteSharp.Core.AST;

public abstract record Stmt : Element
{
    public abstract int GetLine();
    public abstract string GetFile();
    
    public interface INamedStmt : Element
    {
        Token GetNameToken();
    }

    public record Expression(Expr Expr) : Stmt
    {
        public override int GetLine()
        {
            return Expr.GetLine();
        }

        public override string GetFile()
        {
            return Expr.GetFile();
        }

        public override void Accept(IVisitor visitor)
        {
            visitor.VisitExprStmt(this);
        }
    }
    
    public record Function(Token Name, FunctionType Type, List<Param> Params, List<Stmt> Body, FunctionModifier Modifier, Type ReturnType, List<Param> TypeParams) : Stmt, INamedStmt
    {
        public override int GetLine()
        {
            return Name.Line;
        }

        public override string GetFile()
        {
            return Name.File;
        }

        public Token GetNameToken()
        {
            return Name;
        }

        public override void Accept(IVisitor visitor)
        {
            visitor.VisitFunctionStmt(this);
        }
    }

    public record Var(Token Name, Type Type, Expr? Initializer, FieldModifier Modifier) : Stmt, INamedStmt
    {
        public override int GetLine()
        {
            return Name.Line;
        }

        public override string GetFile()
        {
            return Name.File;
        }

        public Token GetNameToken()
        {
            return Name;
        }

        public override void Accept(IVisitor visitor)
        {
            visitor.VisitVarStmt(this);
        }
    }
    
    public record Block(List<Stmt> Statements, int LineNumber, string CurrentFile) : Stmt, INamedStmt
    {
        public override int GetLine()
        {
            return LineNumber;
        }

        public override string GetFile()
        {
            return CurrentFile;
        }

        public Token GetNameToken()
        {
            return Token.Identifier("<block>", GetFile());
        }

        public override void Accept(IVisitor visitor)
        {
            visitor.VisitBlockStmt(this);
        }
    }
    
    public record If(Expr Condition, Stmt ThenBranch, Stmt? ElseBranch) : Stmt
    {
        public override int GetLine()
        {
            return Condition.GetLine();
        }

        public override string GetFile()
        {
            return Condition.GetFile();
        }

        public override void Accept(IVisitor visitor)
        {
            visitor.VisitIfStmt(this);
        }
    }

    public record While(Expr Condition, Stmt Body) : Stmt
    {
        public override int GetLine()
        {
            return Condition.GetLine();
        }

        public override string GetFile()
        {
            return Condition.GetFile();
        }

        public override void Accept(IVisitor visitor)
        {
            visitor.VisitWhileStmt(this);
        }
    }

    public record Break(Token Keyword) : Stmt
    {
        public override int GetLine()
        {
            return Keyword.Line;
        }

        public override string GetFile()
        {
            return Keyword.File;
        }

        public override void Accept(IVisitor visitor)
        {
            visitor.VisitBreakStmt(this);
        }
    }

    public record Continue(Token Keyword) : Stmt
    {
        public override int GetLine()
        {
            return Keyword.Line;
        }

        public override string GetFile()
        {
            return Keyword.File;
        }

        public override void Accept(IVisitor visitor)
        {
            visitor.VisitContinueStmt(this);
        }
    }
    
    public record Return(Token Keyword, Expr? Value) : Stmt
    {
        public override int GetLine()
        {
            return Keyword.Line;
        }

        public override string GetFile()
        {
            return Keyword.File;
        }

        public override void Accept(IVisitor visitor)
        {
            visitor.VisitReturnStmt(this);
        }
    }

    public record Import(Token Keyword, Token What) : Stmt
    {
        public override int GetLine()
        {
            return Keyword.Line;
        }

        public override string GetFile()
        {
            return Keyword.File;
        }

        public override void Accept(IVisitor visitor)
        {
            visitor.VisitImportStmt(this);
        }
    }

    public record Class(
        Token Name,
        List<Function> Methods,
        List<Var> FieldDefaults,
        Expr.Variable? Superclass,
        List<Expr.Variable> Superinterfaces,
        ClassModifier Modifier,
        List<Param> TypeParams) : Stmt, INamedStmt
    {
        public override int GetLine()
        {
            return Name.Line;
        }

        public override string GetFile()
        {
            return Name.File;
        }

        public Token GetNameToken()
        {
            return Name;
        }

        public override void Accept(IVisitor visitor)
        {
            visitor.VisitClassStmt(this);
        }
    }
    
    public record Interface(Token Name, List<Function> Methods, List<Expr.Variable> Superinterfaces, List<Param> TypeParameters) : Stmt, INamedStmt
    {
        public override int GetLine()
        {
            return Name.Line;
        }

        public override string GetFile()
        {
            return Name.File;
        }

        public Token GetNameToken()
        {
            return Name;
        }

        public override void Accept(IVisitor visitor)
        {
            visitor.VisitInterfaceStmt(this);
        }
    }
    
    public record TryCatch(Token TryToken, Token CatchToken, Block TryBody, Param CatchVariable, Block CatchBody) : Stmt
    {
        public override int GetLine()
        {
            return TryToken.Line;
        }

        public override string GetFile()
        {
            return TryToken.File;
        }

        public override void Accept(IVisitor visitor)
        {
            visitor.VisitTryCatchStmt(this);
        }
    }

    public record Throw(Expr Expr) : Stmt
    {
        public override int GetLine()
        {
            return Expr.GetLine();
        }

        public override string GetFile()
        {
            return Expr.GetFile();
        }

        public override void Accept(IVisitor visitor)
        {
            visitor.VisitThrowStmt(this);
        }
    }
    
    public interface IVisitor
    {
        void VisitExprStmt(Expression stmt);
        void VisitVarStmt(Var stmt);
        void VisitBlockStmt(Block stmt);
        void VisitIfStmt(If stmt);
        void VisitWhileStmt(While stmt);
        void VisitBreakStmt(Break stmt);
        void VisitContinueStmt(Continue stmt);
        void VisitFunctionStmt(Function stmt);
        void VisitReturnStmt(Return stmt);
        void VisitClassStmt(Class stmt);
        void VisitInterfaceStmt(Interface stmt);
        void VisitImportStmt(Import stmt);
        void VisitTryCatchStmt(TryCatch stmt);
        void VisitThrowStmt(Throw stmt);
    }
    
    public abstract void Accept(IVisitor visitor);
}